---   
 <img align="left" width="75" height="75"  src="https://upload.wikimedia.org/wikipedia/en/c/c8/University_of_the_Punjab_logo.png"> 

<h1 align="center">Department of Data Science</h1>
<h1 align="center">Course: Tools and Techniques for Data Science</h1>

---
<h3><div align="right">Instructor: Muhammad Arif Butt, Ph.D.</div></h3>    

<h1 align="center">Lecture 2.2</h1>

## _01-hellojupyter.ipynb_

### [Click to read about Jupyter Notebook](https://jupyter.readthedocs.io/en/latest/content-quickstart.html)

## A review of the Jupyter notebook Interface
IPython Notebooks are just files (.ipynb) on your file system, containing JSON data
1. **Header:**  At the top of the notebook document is the header which contains the _notebook title_, _menu bar_, and _toolbar_, which are used to control notebook navigation and document structure.
2. **Body:** The body of a notebook is composed of cells. Each cell contains either markdown, code input, code output, or raw text.
3. **Types of cells:**
    - **Markdown cells** are used to build a nicely formatted narrative around the code in the document. These cells can either be rendered or unrendered. When they are rendered, a nice formatted representation of the cell's contents will be presented. When they are unrendered, the raw text source of the cell will be presented.
    - **Code cells**  are used to define the computational code in the document. They come in two forms:
        - **Input cell** where the user types the code to be executed, and  
        - **output cell** which is the representation of the executed code. Depending on the code, this representation may be a simple scalar value, or something more complex like a plot or an interactive widget.
    -**Raw cells:** These are used when text needs to be included in raw form, without execution or transformation.
4. **Modes of a cell:**
    - **Edit mode** is indicated by a **green** cell border and a prompt showing in the editor area. When a cell is in edit mode, you can type into the cell, like a normal text editor. Enter edit mode by pressing `Enter` or using the mouse to click on a cell's editor area.
    - **Command mode** is indicated by a **blue** cell border. When in command mode, the structure of the notebook can be modified as a whole, but the text in individual cells cannot be changed. Most importantly, the keyboard is mapped to a set of shortcuts for efficiently performing notebook and cell actions. Enter command mode by pressing `Esc` or using the mouse to click *outside* a cell's editor area.

5. **Keyboard Short cuts:** Keyboard shortcuts are a very popular aspect of the Jupyter environment because they facilitate a speedy cell-based workflow. Following shortcuts work in command mode:
    - Press H to see the list of short cuts
    - Shift+Enter to execute code of a code-cell and to render the contents of a markdown cell. It also shifts control to cell down. If there is none it creates one. To avoid shifting control to lower cell after you execute a cell code use <Ctrl+Enter> instead
    - Scroll up and down your cells with your Up and Down keys.
    - Press A or B to insert a new cell above or below the active cell.
    - DD (D twice) will delete the active cell.
    - L to toggle view linenumbers within a cell
    - Z will undo cell deletion. 
    - M will transform the active cell to a Markdown cell.
    - Y will set the active cell to a code cell.
    - <Command+/> to toggle between commenting and uncommenting the current LOC or highlighted LOCs
    - Hold Shift and press Up or Down to select multiple cells at once. With multiple cells selected, Shift + M will merge your selection.

## 1. Performing Arithmetic Operations using Python

Let's begin by using Python as a calculator. You can write and execute Python using a code cell within Jupyter. 

> **Working with cells**: To create a new cell within Jupyter, you can select "Insert > Insert Cell Below" from the menu bar or just press the "+" button on the toolbar. You can also use the keyboard shortcut `Esc+B` to create a new cell. Once a cell is created, click on it to select it. You can then change the cell type to code or markdown (text) using the "Cell > Cell Type" menu option. You can also use the keyboard shortcuts `Esc+Y` and `Esc+M`. Double-click a cell to edit the content within the cell. To apply your changes and run a cell, use the "Cell > Run Cells" menu option or click the "Run" button on the toolbar or just use the keyboard shortcut `Shift+Enter`. You can see a full list of keyboard shortcuts using the "Help > Keyboard Shortcuts" menu option.

Run the code cells below to perform calculations and view their result. Try changing the numbers and run the modified cells again to see updated results.

In [25]:
2 + 3 + 9

14

In [26]:
99 - 73

26

In [27]:
23.54 * -1432

-33709.28

In [28]:
100 / 7

14.285714285714286

In [29]:
100 // 7

14

In [30]:
100 % 7

2

In [31]:
5 ** 3

125

As you might expect, operators like `/` and `*` take precedence over other operators like `+` and `-` as per mathematical conventions. You can use parentheses, i.e. `(` and `)`, to specify the order in which operations are performed.

In [32]:
((2 + 5) * (17 - 3)) / (4 ** 3)

1.53125

Python supports the following arithmetic operators:

| Operator   | Purpose           | Example     | Result    |
|------------|-------------------|-------------|-----------|
| `+`        | Addition          | `2 + 3`     | `5`       |
| `-`        | Subtraction       | `3 - 2`     | `1`       |
| `*`        | Multiplication    | `8 * 12`    | `96`      |
| `/`        | Division          | `100 / 7`   | `14.28..` |
| `//`       | Floor Division    | `100 // 7`  | `14`      |    
| `%`        | Modulus/Remainder | `100 % 7`   | `2`       |
| `**`       | Exponent          | `5 ** 3`    | `125`     |


Beginner programmers should try solving some simple problems from this page:
https://www.math-only-math.com/worksheet-on-word-problems-on-four-operations.html 

## 2. Use of Comments
Note that we're using the `#` character to add *comments* within our code. 

> **Comments**: Comments and blank lines are ignored during execution, but they are useful for providing information to humans (including yourself) about what the code does. Comments can be inline (at the end of some code), on a separate line, or even span multiple lines. 

Inline and single-line comments start with `#`, whereas multi-line comments begin and end with three quotes, i.e. `"""`. Here are some examples of code comments:

In [9]:
# This is a single line comment

"""
A multi-line comment starts and ends with a tripple double quotes
A docstring is a documentation string. As a comment, this Python Syntax is used to explain code.
But unlike comments, they are more specific. Also, they are retained at runtime.
This way, the programmer can inspect them at runtime. Delimit a docstring using three double-quotes. 
"""

'''
A multi-line comment
starts and ends with a tripple single quotes
'''
x=10  # ends with newline
x

10

## 3. Use of print() statement
> **`print`**: The `print` is a Python built-in function that used to display information. It takes one or more inputs, which can be text (within quotes, e.g., `"this is some text"`), numbers, variables, mathematical expressions, etc. We'll learn more about variables the next tutorial.

> Jupyter allows you to write multiple lines of code within a single code cell. However, the result of the last line of code within a cell is displayed as the output. 


In [10]:
# Note only the last command appears in output
a=10
b=a-5
c=a+b
a
b
c

15

In [11]:
# Therefore better to use print() statement
a=10
b=a-5
c=a+b
print("a = ", a)
print("b = ", b)
print("c = ", c)
# You can also fit in more than one statement on one line by separating them with a semicolon.
d = 5 + 3; print('d = ', d)


a =  10
b =  5
c =  15
d =  8


## 4. Getting Help in IPython
Within IPython you have various way to access help:
-  `?`          Introduction and overview of IPython's features (this screen).
-  `help`       Access Python's own help system.
-  `object?`    Details about 'object'.
-  `object??`   More detailed, verbose information about 'object'.
-  `%quickref`  Quick reference of all IPython specific syntax and magics.

In [12]:
?

In [13]:
?print

In [14]:
??print

In [15]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



In [16]:
# Get help about "keywords", "symbols", "topics", ...
# Type "quit" to exit this interactive command
help()


Welcome to Python 3.8's help utility!

If this is your first time using Python, you should definitely check out
the tutorial on the Internet at https://docs.python.org/3.8/tutorial/.

Enter the name of any module, keyword, or topic to get help on writing
Python programs and using Python modules.  To quit this help utility and
return to the interpreter, just type "quit".

To get a list of available modules, keywords, symbols, or topics, type
"modules", "keywords", "symbols", or "topics".  Each module also comes
with a one-line summary of what it does; to list the modules whose name
or summary contain a given string such as "spam", type "modules spam".

help> q

You are now leaving help and returning to the Python interpreter.
If you want to ask for help on a particular object directly from the
interpreter, you can type "help(object)".  Executing "help('string')"
has the same effect as typing a particular string at the help> prompt.


In [17]:
msg = "Learning is fun with Arif Butt"
date = "October 15, 2021"
msg
print(msg)
print("Today's message is: ", msg)
print("Message of arif is ", msg, "and today's date is ", date)

Learning is fun with Arif Butt
Today's message is:  Learning is fun with Arif Butt
Message of arif is  Learning is fun with Arif Butt and today's date is  October 15, 2021


## 5. Running bash shell commands

In [18]:
!pwd
!date
!touch f1.txt
! ls

/Users/arif/Documents/0-DS-522/Demo-Files-Repo/Section-2 (Basics of Python Programming)/Lec-2.02 (Anaconda and Jupyter Notebook)
Tue Nov 30 19:41:08 PKT 2021
01-hellojupyter.ipynb     f1.txt
02-markdown-example.ipynb images


## 6. Magic Commands
- The magic function system provides a series of functions which allow you to control the behavior of IPython itself, plus a lot of system-type features. There are two kinds of magics, line-oriented and cell-oriented.
    - Line magics are prefixed with the % character and work much like OS command-line calls: they get as an argument the rest of the line, where arguments are passed without parentheses or quotes
    - Cell magics are prefixed with a double %%, and they are functions that get as an argument not only the rest of the line, but also the lines below it in a separate argument.  These magics are called with two arguments: the rest of the call line and the body of the cell, consisting of the lines below the first.

In [19]:
%quickref

In [20]:
%magic

In [21]:
%lsmagic

Available line magics:
%alias  %alias_magic  %autoawait  %autocall  %automagic  %autosave  %bookmark  %cat  %cd  %clear  %colors  %conda  %config  %connect_info  %cp  %debug  %dhist  %dirs  %doctest_mode  %ed  %edit  %env  %gui  %hist  %history  %killbgscripts  %ldir  %less  %lf  %lk  %ll  %load  %load_ext  %loadpy  %logoff  %logon  %logstart  %logstate  %logstop  %ls  %lsmagic  %lx  %macro  %magic  %man  %matplotlib  %mkdir  %more  %mv  %notebook  %page  %pastebin  %pdb  %pdef  %pdoc  %pfile  %pinfo  %pinfo2  %pip  %popd  %pprint  %precision  %prun  %psearch  %psource  %pushd  %pwd  %pycat  %pylab  %qtconsole  %quickref  %recall  %rehashx  %reload_ext  %rep  %rerun  %reset  %reset_selective  %rm  %rmdir  %run  %save  %sc  %set_env  %store  %sx  %system  %tb  %time  %timeit  %unalias  %unload_ext  %who  %who_ls  %whos  %xdel  %xmode

Available cell magics:
%%!  %%HTML  %%SVG  %%bash  %%capture  %%debug  %%file  %%html  %%javascript  %%js  %%latex  %%markdown  %%perl  %%prun  %%pypy  %%

In [22]:
%history

2 + 3 + 9
99 - 73
23.54 * -1432
100 / 7
100 // 7
100 % 7
5 ** 3
((2 + 5) * (17 - 3)) / (4 ** 3)
# This is a single line comment

"""
A multi-line comment starts and ends with a tripple double quotes
A docstring is a documentation string. As a comment, this Python Syntax is used to explain code.
But unlike comments, they are more specific. Also, they are retained at runtime.
This way, the programmer can inspect them at runtime. Delimit a docstring using three double-quotes. 
"""

'''
A multi-line comment
starts and ends with a tripple single quotes
'''
x=10  # ends with newline
x
# Note only the last command appears in output
a=10
b=a-5
c=a+b
a
b
c
# Therefore better to use print() statement
a=10
b=a-5
c=a+b
print("a = ", a)
print("b = ", b)
print("c = ", c)
# You can also fit in more than one statement on one line by separating them with a semicolon.
d = 5 + 3; print('d = ', d)
?
?print
??print
help(print)
# Get help about "keywords", "symbols", "topics", ...
# Type "quit" to exit this

In [23]:
%timeit range(1000)

243 ns ± 10.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [24]:
%%bash
echo "My shell is:" $SHELL
echo "User:" $USER

My shell is: /bin/bash
User: arif


## Check your Concepts

Try answering the following questions to test your understanding of the topics covered in this notebook:

1. What is a Jupyter notebook? 
2. How do you add a new code cell below an existing cell?
3. How do you add a new Markdown cell below an existing cell?
4. How do you convert a code cell to a Markdown cell or vice versa?
5. How do you execute a code cell within Jupyter?
6. What is Markdown? Why is it useful?
7. How do you create headings of different sizes using Markdown?
8. How do you create bulleted and numbered lists using Markdown?
9. How do you create bold or italic text using Markdown?
10. How do you include links & images within Markdown cells?
11. How do you include code blocks within Markdown cells?
12. Is it possible to execute the code blocks within Markdown cells?
13. How to run operating system commands from a notebook cell?
14. Mention three different ways to get help about built-in Python functions from a notebook cell? 
15. What different arithmetic operators are available in Python?
16. Explore different line and cell magic functions of IPython